server.py

```diff
@@ -117,61 +149,44 @@ def server_static(filepath):
 
 @app.get("/ranking/<code>")
 def contest_ranking(code):
-    with shelve.open(database_path) as submission_record:
-        order = [
-            (
-                user,
-                len(
-                    set(
-                        [
-                            attempt.question
-                            for attempt in submissions
-                            if (
-                                attempt.is_correct
-                                and (int(attempt.question) in contests[code].questions)
-                                and attempt.contest == code
-                                and attempt.time <= contests[code].end_time
-                                and attempt.time >= contests[code].start_time
-                            )
-                        ]
-                    )
-                ),
-            )
-            for user, submissions in submission_record.items()
-        ]
-    order.sort(key=lambda x: x[1], reverse=True)
-    order = [entry for entry in order if entry[1] > 0]
-    order = [(user, score, rank) for rank, (user, score) in enumerate(order, start=1)]
+    order = (
+        Submission.select(
+            Submission.username, fn.count(Submission.question).alias("score")
+        )
+        .where((Submission.is_correct == True) & (Submission.contest == code))
+        .group_by(Submission.username)
+        .order_by(fn.count(Submission.question).desc())
+    )
+    order = [(x.username, x.score) for x in order]
```

Comment:
```suggestion
    order =list(order.tuples())
```

Since you've already "selected" certain parts you can directly turn them into tuples or dicts